templates/components/common/polyfill-script.html

```diff
@@ -0,0 +1,32 @@
+<script>
+    {{!--
+        Check for modern browser features, and load polyfills if browser does not appear to support features
+        we need.
+
+            Polyfill approach is all or nothing - please import needed polyfills in /assets/js/polyfills.js
+
+        Inspired by https://philipwalton.com/articles/loading-polyfills-only-when-needed/
+            --}}
+    function browserSupportsAllFeatures() {
+        return window.Promise
+            && window.fetch
+            && window.URL
+            && window.URLSearchParams
+            && window.WeakMap
+            // object-fit support
+            && ('objectFit' in document.documentElement.style);
+    }
+
+    function loadScript(src) {
+        var js = document.createElement('script');
+        js.src = src;
+        js.onerror = function () {
+            console.error('Failed to load polyfill script ' + src);
+        };
+        document.head.appendChild(js);
+    }
+
+    if (!browserSupportsAllFeatures()) {
```

Comment:
Yep, I came across this approach of generating 2 different bundles in my research as well.

I wonder if it complicates testing - can I have confidence that two different bundles will behave the same?

Reply:
To make this "magical", we'd need something like this I think: https://github.com/babel/babel/issues/6625

Reply:
Thanks @bookernath. I thought `babel/preset-env`'s `useBuiltIns: usage` setting [already does that](https://babeljs.io/docs/en/babel-preset-env#usebuiltins-usage). But it does the selective injections of polyfills at compile time, as opposed to runtime like `polyfill.io`.

So my suggestion basically utilises that feature to produce two sets of bundles targeting two different ranges of browser versions in build time, and use a user agent check in runtime to determine which one to load. So there's little magic to it. That separation could potentially extend to the transpiled code as well if you can work out a way to load it, not just polyfills (as different browsers require different levels of transpilation). If you just want to focus on reducing the size of polyfills, then the end result might be less effective than what you currently have because you're not including anything. But in exchange, you will have a more reliable check. I'm not suggesting we should, in runtime, work out what features we have and get back the minimum required polyfills for a specific browser version. For that, you'll need a service like `polyfill.io` which I'm also against.

In terms of testing, I don't think it'd create extra work in comparison to other conditional loading approaches, because we'll just be relying on `babel/preset-env` to figure out the browser differences and respond to changes in the codebase. It'd be the same as what we have right now. But the difference is that there will be a runtime check, which has to be done anyway if we go with other conditional loading approaches.

Anyway, it's just an idea that I have that might be able to address the concerns that I have listed above. I haven't tested myself so it might not address any or even create more problems lol.

Reply:
I was just talking to @icatalina about this. He suggested maybe instead of doing a user agent check, we could just use `module` / `nomodule` to target new and old browsers. The `target` can be set to [esmodules: true](https://babeljs.io/docs/en/babel-preset-env#targetsesmodules) for the former, and it can be set as false (or the same `last 2 versions` target we currently have) for the latter. The idea is based on [this article](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/), which is written by the same person a year later. There are some quirks so we might need to tweak a few things in order to get it to work. Anyway, I'm going to see if I can try something similar in checkout.

Reply:
I saw this article as well... it's super interesting, but it seems like sort of a "moment in time" one-off approach, as opposed to something more generic. I'm not against using it though! We're mostly just talking about pesky IE11 these days anyways ;)

Reply:
It is a moment in time approach, but that moment includes most of the things that require polyfills. 

> Every browser that supports <script type="module"> also supports async/await
> Every browser that supports <script type="module"> also supports Classes.
> Every browser that supports <script type="module"> also supports arrow functions.
> Every browser that supports <script type="module"> also supports fetch, and Promises, and Map, and Set, and much more! 

And as you said, it is mainly IE, any other browser tends to auto-update and you can assume to have to support only 2 or 3 versions.

The beauty of the proposed solution is that everything loads directly and natively in the browser, fully utilising caching, preloading, preprocessing and whatnot.